File: 2017/AoC17-10ab.swift
postfix operator -

extension Int {
  static postfix func - (n: Int) -> Int {
    return n - 1
  }
}

extension Array {
  func chunks(of size: Int) -> [[Element]] {
    return self.indices.compactMap {
      $0 % size == 0 ? Array(self[$0..<Swift.min($0+size, self.count)]) : nil
    }
  }
}

func part1(_ input: String) -> Int {
  let dirs = input.split(separator: ",").map { Int($0)! }

  let len = 256
  var nums = Array(0..<256)

  var skip = 0
  var i = 0
  for n in dirs {
    let r = (i..<i+n).map { $0 & len- }
    for (w, x) in zip(r.prefix(n/2), r.reversed()) {
      nums.swapAt(w, x)
    }

    i = i + n + skip
    skip += 1
  }

  return nums[0] * nums[1]
}

func knotHash(_ input: String) -> String {
  let dirs: [Int] = input.map { Int($0.asciiValue!) } + [17, 31, 73, 47, 23]

  let len = 256
  var nums = Array(0..<256)

  var skip = 0
  var i = 0
  for _ in 1...64 {
    for n in dirs {
      let r = (i..<i+n).map { $0 & len- }
      for (w, x) in zip(r.prefix(n/2), r.reversed()) {
        nums.swapAt(w, x)
      }

      i = i + n + skip
      skip += 1
    }
  }

  return nums.chunks(of: 16).reduce("") {
    let n = $1.reduce(0, ^)
    return $0 + (n < 0x10 ? "0" : "") + String(n, radix: 16)
  }
}

let input = readLine()!

print("Part 1 answer:", part1(input))
print("Part 2 answer:", knotHash(input))

File: 2017/AoC17-10ab.rs
use std::fs;
use std::ops::BitXor;

const LEN: usize = 256;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  println!("Part 1 answer: {}", part1(&input));
  println!("Part 2 answer: {}", knot_hash(&input));
}

fn part1(input: &str) -> u32 {
  let dirs = input.split(',').map(|chunk| chunk.parse::<usize>().unwrap());

  let mut nums: [u32; LEN] = core::array::from_fn(|i| i as u32);

  let mut skip: usize = 0;
  let mut i: usize = 0;
  for n in dirs {
    let a = (i..i+n).map(|x| x % LEN);
    for (w, x) in a.to_owned().rev().zip(a).take(n/2) {
      nums.swap(w, x);
    }

    i = i + n + skip;
    skip += 1;
  }

  nums[0] * nums[1]
}

fn knot_hash(input: &str) -> String {
  let dirs: Vec<usize> = input.as_bytes().iter()
    .map(|&n| n as usize)
    .chain([17usize, 31, 73, 47, 23])
    .collect();

  let mut nums: [u32; LEN] = core::array::from_fn(|i| i as u32);

  let mut skip: usize = 0;
  let mut i: usize = 0;
  for _ in 1..=64 {
    for &n in &dirs {
      let a = (i..i+n).map(|x| x % LEN);
        for (w, x) in a.to_owned().rev().zip(a).take(n/2) {
        nums.swap(w, x);
      }

      i = i + n + skip;
      skip += 1;
    }
  }

  nums.chunks(16).fold(String::new(), |acc, chunk| {
    let n = chunk.iter().fold(0, u32::bitxor);
    acc + &format!("{n:0>2x}")
  })
}